modules/carbonAnalyticsAdapter.js

```diff
@@ -0,0 +1,241 @@
+      case CONSTANTS.EVENTS.TCF2_ENFORCEMENT: {
+        registerEngagement();
+        let event = createBaseEngagementEvent(args);
+
+        event.tcf_events = args; // Do we only want relevant information here?
+
+        sendEngagementEvent(event, 'tcf_enforcement');
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+  }
+});
+
+// save the base class function
+carbonAdapter.originEnableAnalytics = carbonAdapter.enableAnalytics;
+
+// override enableAnalytics so we can get access to the config passed in from the page
+carbonAdapter.enableAnalytics = function (config) {
+  if (config?.options?.parentId) {
+    parentId = config.options.parentId;
+  } else {
+    logError('required config value "parentId" not provided');
+  }
+
+  if (config?.options.endpoint) {
+    analyticsHost = config.options.endpoint;
+  } else {
+    logError('required config value "endpoint" not provided');
+  }
+
+  pageViewId = generateUUID();
+  profileId = getProfileId();
+  sessionId = getSessionId();
+
+  pageEngagement = { // create the initial page engagement event
+    ttl: 60 * 1, // should be multiple of 60, unit is seconds
+    count: 0,
+    id: generateUUID(),
+    timeLastEngage: Date.now()
+  };
+
+  let event = createBaseEngagementEvent()
+  sendEngagementEvent(event, 'page_load');
+
+  carbonAdapter.originEnableAnalytics(config); // call the base class function
+};
+
+function getProfileId() {
+  if (storage.localStorageIsEnabled()) {
+    let localStorageId = storage.getDataFromLocalStorage(PROFILE_ID_KEY);
+    if (localStorageId && localStorageId != '') {
+      if (storage.cookiesAreEnabled()) {
+        storage.setCookie(PROFILE_ID_COOKIE, localStorageId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+      }
+
+      return localStorageId;
+    }
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(PROFILE_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      if (storage.localStorageIsEnabled()) {
+        storage.setDataInLocalStorage(PROFILE_ID_KEY, cookieId);
+      }
+
+      storage.setCookie(PROFILE_ID_COOKIE, cookieId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+
+      return cookieId;
+    }
+  }
+
+  let newId = generateUUID();
+  if (storage.localStorageIsEnabled()) {
+    storage.setDataInLocalStorage(PROFILE_ID_KEY, newId);
+  }
+
+  if (storage.cookiesAreEnabled()) {
+    storage.setCookie(PROFILE_ID_COOKIE, newId, new Date(Date.now() + 89 * DAY_MS), 'Lax');
+  }
+
+  return newId;
+}
+
+function getSessionId() {
+  if (storage.cookiesAreEnabled()) {
+    let cookieId = storage.getCookie(SESSION_ID_COOKIE);
+    if (cookieId && cookieId != '') {
+      storage.setCookie(SESSION_ID_COOKIE, cookieId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+
+      return cookieId;
+    }
+  }
+
+  let newId = generateUUID();
+
+  if (storage.cookiesAreEnabled()) {
+    storage.setCookie(SESSION_ID_COOKIE, newId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
+  }
+
+  return newId;
+}
```

Comment:
If we have no cookies we still need a Session ID as it's not an optional field, it would just be a session that exists on only this page visit.

You are right that with a small refactor we can remove the need to check for cookies being enabled twice, I'm thinking something along the lines of: ```function getSessionId() {
  if (storage.cookiesAreEnabled()) {
    let cookieId = storage.getCookie(SESSION_ID_COOKIE);
    if (cookieId && cookieId != '') {
      storage.setCookie(SESSION_ID_COOKIE, cookieId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');

      return cookieId;
    }

    let newId = generateUUID();
    storage.setCookie(SESSION_ID_COOKIE, newId, new Date(Date.now() + 5 * MINUTE_MS), 'Lax');
  }

  return generateUUID();
}```